core/market/decentralized/src/negotiation/requestor.rs

```diff
@@ -194,15 +214,113 @@ impl RequestorBroker {
         Ok(id)
     }
 
+    pub async fn wait_for_approval(
+        &self,
+        id: &AgreementId,
+        timeout: f32,
+    ) -> Result<ApprovalStatus, WaitForApprovalError> {
+        // TODO: Check if we are owner of Proposal
+        // TODO: What to do with 2 simultaneous calls to wait_for_approval??
+        //  should we reject one? And if so, how to discover, that to calls were made?
+        // TODO: Do we have race conditions here, between checking Agreement state and
+        //  waiting for notifications? Event could come, before we start waiting and
+        //  it will be lost.
+        let timeout = Duration::from_secs_f32(timeout.max(0.0));
+        let mut notifier = self.agreement_notifier.listen(id);
+
+        // Loop will wait for events notifications only one time. It doesn't have to be loop at all,
+        // but it spares us doubled getting agreement and mapping statuses to return results.
+        // So I think this simplification is worth confusion, that it cause.
+        loop {
+            let agreement = self
+                .db()
+                .as_dao::<AgreementDao>()
+                .get(id)
+                .await
+                .map_err(|e| WaitForApprovalError::FailedGetFromDb(id.clone(), e))?;
+
+            match agreement.state {
+                AgreementState::Approved => return Ok(ApprovalStatus::Ok),
+                AgreementState::Rejected => return Ok(ApprovalStatus::Rejected),
+                AgreementState::Cancelled => return Ok(ApprovalStatus::Cancelled),
+                AgreementState::Expired => Err(WaitForApprovalError::AgreementExpired(id.clone()))?,
```

Comment:
`return` would be much more readable here than `?` because we have three `return`s above.